## Data Analysis I

Lastly, we will answer the following questions using pandas methods: 

* What are the most popular colors by season?
* What is the most popular clothing item by season?
* What is the effect of promo-codes on the dollar-amount of purchases?
* When do users leave a review?
* Do states vary in the items they purchase in the Fall?

Utilize the documentation provided in each code-block. When you are done with this section of the project, validate that your output matches the screenshot provided in the `docs/part3.md` file and answer the questions located underneath `Data Analysis II` in your own words.

In [1]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

from scipy import stats

In [2]:
# TODO: load `data/processed/shopping_cleaned.csv` as a pandas dataframe

df=pd.read_csv('C:/Users/deema/Desktop/Lab4/shopping-behavior/code/shopping_cleaned.csv')

In [3]:
# TODO: Create a dataframe pivot table where "Season" is your column, "Color" is your index, and "Location" is your value (this value is arbitrary)
# use the len function as the "aggfunc" parameter.
# Save this pivot table to a new variable
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.pivot_table.html

Nu_1_pivot_table = df.pivot_table(index='Color', columns='Season', values='Location', aggfunc=len, fill_value=0)


In [4]:
# TODO: Display this pivot table

print(Nu_1_pivot_table)


Season                Fall  Spring  Summer  Winter
Color                                             
Aubergine              101       4      14     144
Baby blue               17      83      99      24
Black                  103       5      12     155
Brick red              103       7      14     140
Brown                  117       5       5     120
Burnt orange           110       4      10     138
Fuchsia                  5      59     103      21
Lavender                19      63     120      13
Lemon yellow            11      62     112      18
Mauve                  102       6      12     120
Muted mustard yellow   103      10      15     110
Pale peach              13      68     104      15
Periwinkle               9      63     105       9
Ruby Red               112       4      10     132
Terra cotta            108       5      15     131
Turquoise               10      60      96      19
White                   13      73      96      12


In [5]:
# TODO: Select the "Fall" column from this pivot table and display the sort order in descending order
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sort_values.html

sorted_column = Nu_1_pivot_table['Fall'].sort_values(ascending=False)

print(sorted_column)

Color
Brown                   117
Ruby Red                112
Burnt orange            110
Terra cotta             108
Black                   103
Brick red               103
Muted mustard yellow    103
Mauve                   102
Aubergine               101
Lavender                 19
Baby blue                17
White                    13
Pale peach               13
Lemon yellow             11
Turquoise                10
Periwinkle                9
Fuchsia                   5
Name: Fall, dtype: int64


In [6]:
# TODO: Select the "Winter" column from this pivot table and display the sort order in descending order.

sorted_column = Nu_1_pivot_table['Winter'].sort_values(ascending=False)
print(sorted_column)

Color
Black                   155
Aubergine               144
Brick red               140
Burnt orange            138
Ruby Red                132
Terra cotta             131
Brown                   120
Mauve                   120
Muted mustard yellow    110
Baby blue                24
Fuchsia                  21
Turquoise                19
Lemon yellow             18
Pale peach               15
Lavender                 13
White                    12
Periwinkle                9
Name: Winter, dtype: int64


In [7]:
# TODO: Select the "Spring" column from this pivot table and display the sort order in descending order

sorted_column = Nu_1_pivot_table['Spring'].sort_values(ascending=False)
print(sorted_column)

Color
Baby blue               83
White                   73
Pale peach              68
Lavender                63
Periwinkle              63
Lemon yellow            62
Turquoise               60
Fuchsia                 59
Muted mustard yellow    10
Brick red                7
Mauve                    6
Brown                    5
Black                    5
Terra cotta              5
Burnt orange             4
Ruby Red                 4
Aubergine                4
Name: Spring, dtype: int64


In [8]:
# TODO: Select the "Summer" column from this pivot table and display the sort order in descending order

sorted_column = Nu_1_pivot_table['Summer'].sort_values(ascending=False)
print(sorted_column)

Color
Lavender                120
Lemon yellow            112
Periwinkle              105
Pale peach              104
Fuchsia                 103
Baby blue                99
Turquoise                96
White                    96
Muted mustard yellow     15
Terra cotta              15
Brick red                14
Aubergine                14
Mauve                    12
Black                    12
Burnt orange             10
Ruby Red                 10
Brown                     5
Name: Summer, dtype: int64


In [9]:
# TODO: Using the "idmax" method, get the index with the maximum value for each column
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.idxmax.html

Nu_1_pivot_table.idxmax()



Season
Fall          Brown
Spring    Baby blue
Summer     Lavender
Winter        Black
dtype: object

In [10]:
# TODO: Create a dataframe pivot table where "Season" is your column, "Item Purchased" is your index, and "Location" is your value (this value is arbitrary)
# use the len function as the "aggfunc" parameter.
# Save this pivot table to a new variable and display it

Second_pivot_table = df.pivot_table(index='Item Purchased', columns='Season', values='Location', aggfunc=len, fill_value=0)

In [11]:
# TODO: Using the "idmax" method, get the index with the maximum value for each column on the pivot table containing clothing items & season
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.idxmax.html

Second_pivot_table.idxmax()

Season
Fall              Socks
Spring    Running Shoes
Summer           Shorts
Winter         Leggings
dtype: object

In [15]:
# TODO: Create two data-frames for promo-code users and non-promo-code users using boolean indexing
# Save these two dataframes into seperate variables 
# Documentation: https://pandas.pydata.org/docs/user_guide/indexing.html#boolean-indexing
# Hint: Users that have used a promo code have the value "Yes" in the column "Promo Code Used"

#The following code is trying to do:
#Make 2 New DataFrame for the original data
#Using those 2 dataframes create 2 new sets of dataframes that are reading the "Promo Code Used" field
#Evaluating the column to determine "TRUE" or "FALSE"
#That is "Yes" or "No" indicating rather or not the promo code was used.

promo_code_users_df = df[df['Promo Code Used'] == True]
non_promo_code_users_df = df[df['Promo Code Used'] == False]


In [19]:
# TODO: Print the shape of the promo code users dataframe. This should be (1373, 13)

print(promo_code_users_df.shape)


(0, 13)


In [20]:
# TODO: Print the average purchase size of promo code users dataframe. This should be 50.06747998236351

average_purchase_size = promo_code_users_df['Promo Code Used'].mean()
print(average_purchase_size)



nan


In [21]:
# TODO: Print the shape of the non promo code users dataframe. This should be (1785, 13)

print(non_promo_code_users_df.shape)


(0, 13)


In [22]:
# TODO: Print the average purchase size of non promo code users dataframe. This should be 30.155750765104113

average_purchase_size = non_promo_code_users_df['Promo Code Used'].mean()
print(average_purchase_size)


nan


Formulate a null hypothesis regarding the dollar amount spent between promo-code users & non-promo-code users

One null hypothesis could be that the total dollars that those customer with a promo code (promo code user) would spend more money on average versus the customer that did not have a promo code (non-promo code user).

Formulate an alternative hypothesis regarding the dollar amount spent between promo-code users & non-promo-code users

Alternately, if the customer did have a promo code, it may only be applicable to a certain amount of money that the customer will have to spend in order to even have the ability to use the promo code. e.g a promo code will be applicable for all orders OVER $ 200.00. Therefore, if the customer is unable to afford to purchase that amount of merchandise totalling over $ 200.00 they will be automatically restricted from using the promo code.

In [ ]:
# UNABLE TO EXECUTE
# TODO: Using the "ttest_ind" method from "scipy.stats", run a T-Test between the dollar amount spent of promo-code users and non promo-code users
# The t-statistic you should observe is 93.2464714350831
# Documentation: https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html

#ttest for 1.) dollar amount spent of promo-code users and 2.) non promo-code users



In [37]:
#Added my own code for an example

#Calculating the t-value and p-value of 2 indicators A1 and B1

from scipy import stats

# Assuming that A1 and B1 are your data
A1 = [2, 18, 16, 3, 9, 12, 14]
B1 = [32, 12, 34, 23, 30, 19, 26]

t_value, p_value = stats.ttest_ind(A1, B1)

print("t-value: ", t_value)
print("p-value: ", p_value)


t-value:  -3.8663000222524073
p-value:  0.002242872583106794


In [38]:
#Calculating the Standard Deviation of a t-test

import numpy as np

#Suggested dataset
data = [10, 13, 11, 15, 20, 14, 11]

# Calculate standard deviation
std_dev = np.std(data)

print("Standard Deviation: ", std_dev)




Standard Deviation:  3.155817433482074


In [42]:
# TODO: Group your dataframe according to the "Review Rating" column. This should produce two groups for the values "Missing" & "Present"
# Save this grouped dataframe into a new variable named "g_review"
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.groupby.html

df=pd.read_csv('C:/Users/deema/Desktop/Lab4/shopping-behavior/code/shopping_cleaned.csv')
g_review = df.groupby('Review Rating')
print(df['Review Rating'].head()) 

0    4.0
1    4.0
2    5.0
3    5.0
4    4.0
Name: Review Rating, dtype: float64


In [44]:
# TODO: Print out the keys of this dataframe for validation. This should output "dict_keys(['Missing', 'Present'])"
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.keys.html

print(g_review.groups.keys())


dict_keys([1.0, 2.0, 4.0, 5.0])


In [45]:
# TODO: calculate the average of the "Purchase Amount (USD)", "Age", and "Previous Purchases" columns in our grouped dataframe
# output the values for display
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.mean.html
# Documentation: https://pandas.pydata.org/docs/getting_started/intro_tutorials/03_subset_data.html

#The average values from the grouped by range for 'Purchase Amount (USD)', 'Age', 'Previous Purchases

average_values = g_review[['Purchase Amount (USD)', 'Age', 'Previous Purchases']].mean()

# Print the average values
print(average_values)

               Purchase Amount (USD)        Age  Previous Purchases
Review Rating                                                      
1.0                        27.176400  31.933333            5.600000
2.0                        30.879706  32.969697            7.757576
4.0                        30.394100  35.741036            8.000000
5.0                        30.604964  32.891156            8.108844


In [48]:
# TODO: Create a new column called "Prev Purchase Group" on your original dataframe which seperates your dataframe into 
# two groups of shoppers based on their "Previous Purchases" column using the "pd.cut" method. 
# Ensure that you are only creating 2 "bins" and label these respective bins as ["Low", "High"]
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.cut.html

#df_copy=pd.read_csv('C:/Users/deema/Desktop/Lab4/shopping-behavior/code/shopping_cleaned.csv')

#df=pd.read_csv('C:/Users/deema/Desktop/Lab4/shopping-behavior/code/shopping_cleaned.csv')

#df['Prev Purchase Group'] = pd.cut(df['Previous Purchases'], bins=2, labels=['Low', 'High'])

print(df['Prev Purchase Group'].head())

0     Low
1    High
2     Low
3     Low
4     Low
Name: Prev Purchase Group, dtype: category
Categories (2, object): ['Low' < 'High']


In [49]:


# TODO: Using the "value_counts" function, count how many Missing & Present values are in the "Low" group
# Display this value for analysis 
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.Series.value_counts.html
# Hint: You will have to use Boolean Indexing

#Show the 'Low' group
low_group = df[df['Prev Purchase Group'] == 'Low']

#Count the 'Missing' and 'Present' values
value_counts = low_group['Review Rating'].value_counts()

#Print the counts
print(value_counts)

Review Rating
4.0    174
5.0     99
2.0     49
1.0     12
Name: count, dtype: int64


In [50]:
# TODO: Using the "value_counts" function, count how many Missing & Present values are in the "High" group
# Display this value for analysis 
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.Series.value_counts.html
# Hint: You will have to use Boolean Indexing

#Show the 'High' group
high_group = df[df['Prev Purchase Group'] == 'High']

#Count the 'Missing' and 'Present' values
value_counts = high_group['Review Rating'].value_counts()

# Print the counts
print(value_counts)

Review Rating
4.0    77
5.0    48
2.0    17
1.0     3
Name: count, dtype: int64


## Data Analysis II

In the next section, answer the primary analytical questions in your own words.

### Q1

What are the top three colors for Fall & Winter?

Top 3 colors for Fall and Winter are: Fall ( there were some ties so I listed them): 
Fall
Brown                   117
Ruby Red                112
Burnt orange            110

Winter
Black                   155
Aubergine               144
Brick red               140



### Q2

What are the top three colors for Spring & Summer?

Top 3 colors for Spring and Summer are: 
Spring: 
Baby blue               83
White                   73
Pale peach              68


Summer:
Lavender                120
Lemon yellow            112
Periwinkle              105



### Q3

What is the most popular clothing item by season?

Season
Fall              Socks
Spring    Running Shoes
Summer       Sunglasses
Winter         Leggings

### Q4

Observing the p-value that we got when running our t-test between promo-code and non-promo-code users, what can we conclude regarding our null hypothesis?

The Verdict is still out! 

### Q5

Observe the `value_counts` output for the "Review Rating" column for both your "High" and "Low" groups. Proportionally speaking, which group is more likely to leave a review? Why might this be happening from the "human"-perspective? Rationalizations are ok at this point, even if they aren't backed up by data.

I believe that the group most likely to leave a review at all is: Low group. Our dollars count for something, no matter how little we spend. I think that the economy is driving a lot of these factors. Everything is more expensive like 2-3 times as great as it was 2-3 years ago. And even when you are paying less, you are still expecting to have Great Value, Great Product Satisfaction, Great Reliability from your purchase. I would be interested to learn what the ratings were prior to the pandemic and even prior to 2000.